File: mdcc/strategies.rb
module Legostore
  class Strategies
    HOT_TABLE = :hot_table
    HOT_COLUMN = :hot_column
    LOPT = :lopt

    Placement = Struct.new(:column_id, :device_id)

    def self.hot_table_strategy(metadata)
      table_access_counts = metadata.trace.each_with_object(Hash.new(0)) do |entry, counts|
        counts[entry.table_id] += 1
      end

      sorted_tables = table_access_counts.sort_by { |_, count| -count }
      sorted_devices = metadata.devices.values.sort_by { |device| -device.cost_per_gb }

      placements = []
      sorted_tables.each_with_index do |(table_id, _), i|
        table = metadata.get_table(table_id)
        device = sorted_devices[i % sorted_devices.size]
        table.columns.each do |column|
          placements << Placement.new(column.id, device.id)
        end
      end

      placements
    end

    def self.hot_column_strategy(metadata)
      column_access_counts = metadata.trace.each_with_object(Hash.new(0)) do |entry, counts|
        entry.column_ids.each { |col_id| counts[col_id] += 1 }
      end

      sorted_columns = column_access_counts.sort_by { |_, count| -count }
      sorted_devices = metadata.devices.values.sort_by { |device| -device.cost_per_gb }

      sorted_columns.map.with_index do |(col_id, _), i|
        Placement.new(col_id, sorted_devices[i % sorted_devices.size].id)
      end
    end

    def self.lopt_strategy(metadata, budget)
      column_sizes = metadata.columns.transform_values(&:size)
      sorted_columns = column_sizes.sort_by { |_, size| -size }
      sorted_devices = metadata.devices.values.sort_by(&:cost_per_gb)

      placements = []
      total_cost = 0

      sorted_columns.each do |col_id, size|
        sorted_devices.each do |device|
          cost = (size.to_f / (1024 * 1024 * 1024)) * device.cost_per_gb
          if !budget || total_cost + cost <= budget
            placements << Placement.new(col_id, device.id)
            total_cost += cost
            break
          end
        end
      end

      placements
    end

    def self.place_data(strategy, metadata, budget = nil)
      case strategy
      when HOT_TABLE then hot_table_strategy(metadata)
      when HOT_COLUMN then hot_column_strategy(metadata)
      when LOPT then lopt_strategy(metadata, budget)
      else []
      end
    end

    def self.predict_scan_time(placements, table_scan, metadata)
      device_times = Hash.new(0)

      table_scan.column_ids.each do |col_id|
        placement = placements.find { |p| p.column_id == col_id }
        next unless placement

        device = metadata.get_device(placement.device_id)
        device_model = metadata.get_device_model(placement.device_id)
        column = metadata.get_column(col_id)

        if device && device_model && column
          compression_throughput = case device.compression
                                   when COMPRESSION_NONE then device_model.none_throughput
                                   when COMPRESSION_LZ4 then device_model.lz4_throughput
                                   when COMPRESSION_ZSTD then device_model.zstd_throughput
                                   end

          column_time = device_model.seek_time + (column.size.to_f / compression_throughput)
          device_times[placement.device_id] += column_time
        end
      end

      device_times.values.max || 0
    end

    def self.predict_total_time(placements, metadata)
      metadata.trace.sum { |trace| predict_scan_time(placements, trace, metadata) }
    end

    def self.calculate_total_cost(placements, metadata)
      placements.sum do |p|
        device = metadata.get_device(p.device_id)
        column = metadata.get_column(p.column_id)
        if device && column
          device.cost_per_gb * (column.size.to_f / (1024 * 1024 * 1024))
        else
          0
        end
      end
    end

    def self.optimize_placement(budget, metadata)
      placements = place_data(LOPT, metadata, budget)
      total_time = predict_total_time(placements, metadata)
      total_cost = calculate_total_cost(placements, metadata)
      puts "Predicted total time: #{total_time} s"
      puts "Total cost: #{total_cost}"
      placements
    end

    def self.apply_placements(placements, metadata, kv_store, data_retriever)
      placements.each do |placement|
        column = metadata.get_column(placement.column_id)
        target_device = metadata.get_device(placement.device_id)
        if column && target_device
          chunks = data_retriever.column_chunks[placement.column_id] || []
          chunks.each do |chunk|
            if chunk.device_id != placement.device_id
              data_retriever.move_column_chunk(chunk, placement.device_id, metadata, kv_store)
              puts "Moved column #{column.name} to device #{target_device.name}"
            end
          end
        end
      end
    end
  end
end

File: wpaxos/src/strategies.cpp
#include "strategies.hpp"
#include <algorithm>
#include <limits>
#include <iostream>

namespace Mosaic {

std::vector<Placement> Strategies::place_data(PlacementStrategy strategy, Metadata& metadata, float* budget) {
    switch (strategy) {
        case PlacementStrategy::HOT_TABLE:
            return hot_table_strategy(metadata);
        case PlacementStrategy::HOT_COLUMN:
            return hot_column_strategy(metadata);
        case PlacementStrategy::LOPT:
            return lopt_strategy(metadata, budget);
        default:
            return {};
    }
}

std::vector<Placement> Strategies::hot_table_strategy(Metadata& metadata) {
    std::unordered_map<uint32_t, int> table_access_counts;
    for (const auto& entry : metadata.trace) {
        table_access_counts[entry.table_id]++;
    }

    std::vector<std::pair<uint32_t, int>> sorted_tables(table_access_counts.begin(), table_access_counts.end());
    std::sort(sorted_tables.begin(), sorted_tables.end(),
              [](const auto& a, const auto& b) { return a.second > b.second; });

    std::vector<Device*> sorted_devices;
    for (auto& [_, device] : metadata.devices) {
        sorted_devices.push_back(&device);
    }
    std::sort(sorted_devices.begin(), sorted_devices.end(),
              [](const auto& a, const auto& b) { return a->cost_per_gb > b->cost_per_gb; });

    std::vector<Placement> placements;
    for (size_t i = 0; i < sorted_tables.size(); ++i) {
        Table* table = metadata.get_table(sorted_tables[i].first);
        Device* device = sorted_devices[i % sorted_devices.size()];
        for (const auto& column : table->columns) {
            placements.push_back({column.id, device->id});
        }
    }

    return placements;
}

std::vector<Placement> Strategies::hot_column_strategy(Metadata& metadata) {
    std::unordered_map<uint32_t, int> column_access_counts;
    for (const auto& entry : metadata.trace) {
        for (uint32_t col_id : entry.column_ids) {
            column_access_counts[col_id]++;
        }
    }

    std::vector<std::pair<uint32_t, int>> sorted_columns(column_access_counts.begin(), column_access_counts.end());
    std::sort(sorted_columns.begin(), sorted_columns.end(),
              [](const auto& a, const auto& b) { return a.second > b.second; });

    std::vector<Device*> sorted_devices;
    for (auto& [_, device] : metadata.devices) {
        sorted_devices.push_back(&device);
    }
    std::sort(sorted_devices.begin(), sorted_devices.end(),
              [](const auto& a, const auto& b) { return a->cost_per_gb > b->cost_per_gb; });

    std::vector<Placement> placements;
    for (size_t i = 0; i < sorted_columns.size(); ++i) {
        placements.push_back({sorted_columns[i].first, sorted_devices[i % sorted_devices.size()]->id});
    }

    return placements;
}

std::vector<Placement> Strategies::lopt_strategy(Metadata& metadata, float* budget) {
    std::vector<std::pair<uint32_t, uint64_t>> column_sizes;
    for (const auto& [col_id, column] : metadata.columns) {
        column_sizes.emplace_back(col_id, column.size);
    }
    std::sort(column_sizes.begin(), column_sizes.end(),
              [](const auto& a, const auto& b) { return a.second > b.second; });

    std::vector<Device*> sorted_devices;
    for (auto& [_, device] : metadata.devices) {
        sorted_devices.push_back(&device);
    }
    std::sort(sorted_devices.begin(), sorted_devices.end(),
              [](const auto& a, const auto& b) { return a->cost_per_gb < b->cost_per_gb; });

    std::vector<Placement> placements;
    float total_cost = 0.0f;

    for (const auto& [col_id, size] : column_sizes) {
        for (const auto& device : sorted_devices) {
            float cost = (float)size / (1024 * 1024 * 1024) * device->cost_per_gb;
            if (!budget || total_cost + cost <= *budget) {
                placements.push_back({col_id, device->id});
                total_cost += cost;
                break;
            }
        }
    }

    return placements;
}

float Strategies::predict_scan_time(const std::vector<Placement>& placements, const TraceEntry& table_scan, Metadata& metadata) {
    std::unordered_map<uint32_t, float> device_times;

    for (uint32_t col_id : table_scan.column_ids) {
        auto it = std::find_if(placements.begin(), placements.end(),
                               [col_id](const Placement& p) { return p.column_id == col_id; });
        if (it != placements.end()) {
            Device* device = metadata.get_device(it->device_id);
            DeviceModel* device_model = metadata.get_device_model(it->device_id);
            Column* column = metadata.get_column(col_id);

            if (device && device_model && column) {
                float compression_throughput;
                switch (device->compression) {
                    case CompressionAlgorithm::None:
                        compression_throughput = device_model->none_throughput;
                        break;
                    case CompressionAlgorithm::LZ4:
                        compression_throughput = device_model->lz4_throughput;
                        break;
                    case CompressionAlgorithm::ZSTD:
                        compression_throughput = device_model->zstd_throughput;
                        break;
                }

                float column_time = device_model->seek_time + (float)column->size / compression_throughput;
                device_times[it->device_id] += column_time;
            }
        }
    }

    float max_time = 0.0f;
    for (const auto& [_, time] : device_times) {
        max_time = std::max(max_time, time);
    }

    return max_time;
}

float Strategies::predict_total_time(const std::vector<Placement>& placements, Metadata& metadata) {
    float total_time = 0.0f;
    for (const auto& trace : metadata.trace) {
        total_time += predict_scan_time(placements, trace, metadata);
    }
    return total_time;
}

float Strategies::calculate_total_cost(const std::vector<Placement>& placements, Metadata& metadata) {
    float total_cost = 0.0f;
    for (const auto& placement : placements) {
        Device* device = metadata.get_device(placement.device_id);
        Column* column = metadata.get_column(placement.column_id);
        if (device && column) {
            total_cost += device->cost_per_gb * (float)column->size / (1024 * 1024 * 1024);
        }
    }
    return total_cost;
}

std::vector<Placement> Strategies::optimize_placement(float budget, Metadata& metadata) {
    std::vector<Placement> placements = place_data(PlacementStrategy::LOPT, metadata, &budget);
    float total_time = predict_total_time(placements, metadata);
    float total_cost = calculate_total_cost(placements, metadata);
    std::cout << "Predicted total time: " << total_time << " s" << std::endl;
    std::cout << "Total cost: " << total_cost << std::endl;
    return placements;
}

void Strategies::apply_placements(const std::vector<Placement>& placements, Metadata& metadata, KVStore& kv_store, DataRetriever& data_retriever) {
    for (const auto& placement : placements) {
        Column* column = metadata.get_column(placement.column_id);
        Device* target_device = metadata.get_device(placement.device_id);
        if (column && target_device) {
            auto it = data_retriever.column_chunks.find(placement.column_id);
            if (it != data_retriever.column_chunks.end()) {
                for (auto& chunk : it->second) {
                    if (chunk.device_id != placement.device_id) {
                        data_retriever.move_column_chunk(chunk, placement.device_id, metadata, kv_store);
                        std::cout << "Moved column " << column->name << " to device " << target_device->name << std::endl;
                    }
                }
            }
        }
    }
}

} // namespace Mosaic